feat(Worker): add cf-* headers to request

jdanyow opened the pull request:
#17

gja commented on app/worker.js:
```diff
@@ -25,7 +26,7 @@ class Worker {
     this.upstreamHost = upstreamHost;
     this.origin = origin;
-
+    this.ipAddress = ip.address();
```
This is the worker IP address, not the client IP address.

gja commented on app/worker.js:
```diff
@@ -62,9 +63,11 @@ class Worker {
     let response = null;
     let waitUntil = [];
+    const request = new Request(url, { redirect: "manual", ...opts });
+    this.setCFHeaders(request);
```
Make these two a non-instance method? Like buildRequest(url, opts)?

jdanyow replied:
Added a buildRequest method that accepts the express req and a consolidated all the request building logic in server.js. Had to keep the `redirect: 'manual'` behavior in the worker however. What do you think?

jdanyow replied:
I'm wondering if the worker spec should take a dependency on `buildRequest` rather than calling `new Request`, or, would your rather continue building the request in the worker? Maybe `executeFetchEvent` should accept the express request?

gja commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "atob": "^2.1.2",
     "body-parser": "^1.18.3",
     "btoa": "^1.2.1",
+    "ip": "^1.1.5",
```
This shouldn’t be needed. Express gives you the ip on the request

jdanyow commented:
@gja thanks for the feedback. I don't know if I have this quite right yet.  I've added a buildRequest function in server.js and moved all the request building logic in server.js and worker.js to that method. The function takes the express request as it's only argument.

It's a tricky refactor because of the dependency shift. Building the request for the fetch event needs the ip, etc. I'm wondering if the worker spec should take a dependency on `buildRequest` rather than calling `new Request`, or, would your rather I revert and move the request buildinging logic to the worker?

jdanyow commented on app/server.js:
```diff
@@ -1,16 +1,30 @@
-    headers: req.headers,
+function buildRequest(req) {
+  const url = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
+  const ip = req.headers['x-forwarded-for'] || req.connection.remoteAddress.split(':').pop();
```
> `req.connection.remoteAddress.split(':').pop()`

pro: no additional dependencies
con: there are more robust options

gja commented:
Good points. I think it does make sense to keep the logic in the worker instead of in the server. Primarily because the worker should be more reusable.

Is it possible to keep the API signature of worker the same? (taking url and opts). But the Headers are added by the worker. opts.remoteAddress can have the value of node.connection.remoteAddress.

Let me know if you'd like me to pick this up as well, I have a bit of bandwidth tonight.

jdanyow commented:
Thanks @gja, I'm happy to keep plugging away at it but if you have bandwidth and want to pick this up, please go ahead. I think you have a clearer vision for how this should be implemented. I'll probably be able to get back to this in the next few days.
